# BAM-matcher #

A simple tool for determining whether two [BAM files](https://samtools.github.io/hts-specs/SAMv1.pdf) contain reads sequenced from the same sample or patient by counting genotype matches at common SNPs.

Once configuration file is setup, to compare two bam files (sample1.bam and sample2.bam) just run:
```
bam-matcher.py -B1 sample1.bam -B2 sample2.bam
```

which will give an output like:
```

BAM1:	sample1.bam 
BAM2:	sample2.bam 
depth threshold: 15 
____________________________________

Positions with same genotype:   243  
     breakdown:    hom: 51
                   het: 192
____________________________________

Positions with diff genotype:   158 
     breakdown:
                       BAM 1 
               | het  | hom  | subset 
        -------+------+------+------- 
         het   |    0 |    0 |   76  
        -------+------+------+------- 
BAM 2    hom   |    0 |    0 |   -   
        -------+------+------+------- 
         subset|   82 |   -  |   -   
____________________________________

Total sites compared: 401
Fraction of common: 0.605985 (243/401)
CONCLUSION: DIFFERENT SOURCES

```




# Installation #

See the [wiki page](https://bitbucket.org/sacgf/bam-matcher/wiki/Installation) for detailed installation guide.

```
cd /directory/path/where/bam-matcher/is/to/be/installed/
git clone https://bitbucket.org/sacgf/bam-matcher.git
```

The repository includes 3 VCF files which can be used for comparing human data (hg19/GRCh37). 

These VCF files also contain variants extracted from 1000 Genomes project which are all exonic and have high likelihood of switching between REF and ALT alleles (global allele frequency between 0.45 and 0.55). 
The only difference between them is the number of variants contained within.

The repository also includes several BAM files which can be used for testing (under test_data directory), as well as the expected results for various settings.

See the [tutorial](https://bitbucket.org/sacgf/bam-matcher/wiki/Usage) for details on how to test BAM-matcher using the example data.

-----

# Dependencies #

**Python** 

(version 2.7)

**Python libraries**

* PyVCF
* ConfigParser
* Cheetah
* fisher (requires numpy and python-dev libraries).

**Variant Callers**

(Require at least one)

* GATK (requires Java)
* VarScan2 (requires Java and Samtools)
* Freebayes

-----

# Configuration #

BAM-matcher requires a configuration file. If the default configuration file is missing, a template can be generated by the ```--generate-config (-G)``` function. 

```
BAM-matcher.py --generate-config path_to_file_to_be_generated
```

For detailed instruction on how to set up a configuration file, see the [wiki page](https://bitbucket.org/sacgf/bam-matcher/wiki/Configuration).

Most configuration settings can also be overridden at run time.

-----

# Running BAM-matcher #

If configured correctly, to compare two bam files, you just need to run:

```
bam-matcher.py --bam1 sample1.bam --bam2 sample2.bam -o output_report.txt
```

For detailed information on runtime arguments and parameters, see the [wiki page](https://bitbucket.org/sacgf/bam-matcher/wiki/Arguments).

See the [tutorial](https://bitbucket.org/sacgf/bam-matcher/wiki/Usage) on how to test BAM-matcher using the example data.

-----


# LICENCE #

The code is released under the Creative Commons by Attribution licence (http://creativecommons.org/licenses/by/4.0/). You are free to use and modify it for any purpose (including commercial), so long as you cite/link ....




# Who do I talk to? #

Paul (paul.wang @ sa.gov.au)